MdePkg/Library/BasePanicLibSerialPort/BasePanicLibSerialPort.c

```diff
@@ -0,0 +1,68 @@
+/** @file
+  Base Serial Port Panic Library.
+
+  Copyright (c) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Base.h>
+#include <Library/BaseLib.h>
+#include <Library/PrintLib.h>
+#include <Library/SerialPortLib.h>
+#include <Library/PanicLib.h>
+
+//
+// Define the maximum panic message length that this library supports
+//
+#define MAX_PANIC_MESSAGE_LENGTH  0x100
+
+/**
+  Prints a panic message containing a filename, line number, and description.
+  This is always followed by a dead loop.
+
+  Print a message of the form "PANIC <FileName>(<LineNumber>): <Description>\n"
+  to the debug output device.  Immediately after that CpuDeadLoop() is called.
+
+  If FileName is NULL, then a <FileName> string of "(NULL) Filename" is printed.
+  If Description is NULL, then a <Description> string of "(NULL) Description" is printed.
+
+  @param  FileName     The pointer to the name of the source file that generated the panic condition.
+  @param  LineNumber   The line number in the source file that generated the panic condition
+  @param  Description  The pointer to the description of the panic condition.
+
+**/
+VOID
+EFIAPI
+PanicReport (
+  IN CONST CHAR8  *FileName,
+  IN UINTN        LineNumber,
+  IN CONST CHAR8  *Description
+  )
+{
+  CHAR8  Buffer[MAX_PANIC_MESSAGE_LENGTH];
+
+  if (FileName == NULL) {
+    FileName = "(NULL) Filename";
+  }
+
+  if (Description == NULL) {
+    Description = "(NULL) Description";
+  }
+
+  //
+  // Generate the PANIC message in ASCII format
+  //
+  AsciiSPrint (Buffer, sizeof (Buffer), "PANIC [%a] %a(%d): %a\n", gEfiCallerBaseName, FileName, LineNumber, Description);
+
+  // Initialize Serial Port to write the panic message
+  SerialPortInitialize ();
+
+  //
+  // Send the print string to the Logging device device
+  //
+  SerialPortWrite ((UINT8 *)Buffer, AsciiStrLen (Buffer));

```

Comment:
I am not sure which is better among the following two:
1. PanicLib depends on SerialPortLib
2. PanicLib depends on DebugLib

Reply:
Hmm, good question.  Maybe both?

OVMF logs typically go to the qemu debug console.  So it makes sense to have the messages there.  In production VM configs the qemu debug console is typically disabled though, so the messages are discarded.  So reporting serious conditions where the firmware can not continue in a more prominent way such as the serial port makes sense to me.  That already happens with page faults for example.  But I find it a bit irritating to have page faults *not* show up on the qemu debug console.  So logging panics to both serial console and qemu debug console looks best to me.

Reply:
The original idea with the library is that it a platform could easily substitute their own library which implements their desired `PanicReport` function with the serial port version being the most basic one.  I'm good with including the  qemu debug console if that's what people want though.